// osu-api-extended/dist/utility/test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.authorize = exports.login = exports.login_lazer = exports.expired = exports.set_expire = exports.set_token = exports.isLogin = exports.expire = exports.cache_token = void 0;
const request_1 = require("../utility/request");
exports.cache_token = '';
exports.expire = 0;
const isLogin = () => exports.cache_token != '' ? true : false;
exports.isLogin = isLogin;
const set_token = (v) => exports.cache_token = v;
exports.set_token = set_token;
const set_expire = (v) => exports.expire = v;
exports.set_expire = set_expire;
const expired = () => {
    const date = new Date();
    const unix = Math.floor(date.getTime() / 1000);
    console.log('expired', (exports.expire > 0 && unix < exports.expire), date, unix, exports.expire);
    if ((exports.expire > 0 && unix < exports.expire))
        return true;
    return false;
};
exports.expired = expired;
const login_lazer = async (username, password) => {
    if ((0, exports.expired)() && exports.cache_token)
        return;
    const { access_token, expires_in } = await (0, request_1.request)('https://osu.ppy.sh/oauth/token', {
        method: 'POST',
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/json",
        },
        data: JSON.stringify({
            username,
            password,
            grant_type: "password",
            client_id: 5,
            client_secret: 'FGc9GAtyHzeQDshWP5Ah7dega8hJACAJpQtw6OXk',
            scope: "*"
        })
    });
    const date = new Date();
    exports.cache_token = access_token;
    exports.expire = Math.floor(parseFloat(date.setSeconds(date.getSeconds() + expires_in).toString()));
    return { access_token, expires_in };
};
exports.login_lazer = login_lazer;
const login = async (clientId, clientSecret) => {
    if (exports.cache_token)
        return;
    const { access_token, expires_in } = await (0, request_1.request)('https://osu.ppy.sh/oauth/token', {
        method: 'POST',
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/json",
        },
        data: JSON.stringify({
            grant_type: 'client_credentials',
            client_id: clientId,
            client_secret: clientSecret,
            scope: 'public',
            code: 'code',
        })
    });
    exports.cache_token = access_token;
    exports.expire = expires_in;
    return { access_token, expires_in };
};
exports.login = login;
const readline_1 = __importDefault(require("readline"));
const open_1 = __importDefault(require("open"));
const authorize = async (clientId, clientSecret, redirect_uri, scope, state) => {
    if (exports.cache_token)
        return;
    const cl = readline_1.default.createInterface(process.stdin, process.stdout);
    const question = (q) => new Promise((res, rej) => cl.question(q + ': ', (answer) => res(answer)));
    await (0, open_1.default)(`https://osu.ppy.sh/oauth/authorize?client_id=${clientId}&redirect_uri=${redirect_uri}&response_type=code&scope=friends.read%20identify%20public`);
    const code = await question('Paste code here');
    const { access_token, expires_in } = await (0, request_1.request)('https://osu.ppy.sh/oauth/token', {
        method: 'POST',
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/json",
        },
        data: JSON.stringify({
            grant_type: 'authorization_code',
            client_id: clientId,
            client_secret: clientSecret,
            redirect_uri,
            code,
        })
    });
    exports.cache_token = access_token;
    exports.expire = expires_in;
    return { access_token, expires_in };
};
exports.authorize = authorize;
